### ISSUE TITLE
Potential bug in binary mask indexing.

### ISSUE THREAD

**Issue by nkolot** [pytorch/pytorch#32579]
## 🐛 Bug

I came across this issue recently while I was debugging some piece of code. I am not sure whether this is a bug or it is intended behavior. I searched but I did not find something relevant in the docs.

## To Reproduce

The script to reproduce it actually is very easy
```
import torch
x = torch.ones(3, 5)
mask = torch.tensor([0,1,1], dtype=torch.bool)
x1 = x.clone()
x2 = x.clone()
x1[mask] = 0
x2[mask][:, 0] = 0

print(x1)
print(x2)
```

## Expected behavior

I was expecting that the elements x2[1,0] and x2[1,1] to become zero after the operation. Apparently if you apply first the binary mask and then index the result of that it creates a new copy that doesn't share the same memory with the original array.

 - PyTorch Version 1.1:
 - OS : Ubuntu 18.04
 - How you installed PyTorch : pip
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.1 / 7.5
 - GPU models and configuration: RTX 2080Ti

**Comment by ngimel** [comment]
This is intended behavior. x2[mask], if not followed by assignment statement, produces a new tensor that is not sharing storage with original tensor. To do what you want to do you have to index both dimensions in a single indexing operation:
```
import torch
x = torch.randn(3, 5)
mask0 = torch.tensor([0,1,1], dtype=torch.bool)
mask1 = torch.tensor([1,0,0,0,0], dtype=torch.bool)
x[mask0,mask1] = 0
print(x)
```

**Comment by nkolot** [comment]
Thanks!

**Comment by ngimel** [comment]
Or, alternatively, if your real problem is amenable to slicing, you can use slicing, it is more efficient:
```
x[1:,0] = 0
```
slicing is much less flexible than indexing, though.
